src/optionals/log.c

```diff
@@ -0,0 +1,298 @@
+    struct tm *ptm = localtime(&rawtime);
+    if (ptm == NULL) { 
+        return NULL;
+    }
+
+    return ptm;
+}
+
+static Value printLog(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "print() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    struct tm *ptm = timeNow();
+
+    char *msg = AS_CSTRING(args[0]);
+    printf("%04d/%02d/%02d %02d:%02d:%02d %s", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+
+    return newResultSuccess(vm, NIL_VAL);
+}
+
+static Value printlnLog(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "println() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+    
+    struct tm *ptm = timeNow();
+
+    char *msg = AS_CSTRING(args[0]);
+    printf("%04d/%02d/%02d %02d:%02d:%02d %s\n", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+
+    return newResultSuccess(vm, NIL_VAL);
+}
+
+static Value fatalLog(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "fatal() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+    
+    struct tm *ptm = timeNow();
+
+    char *msg = AS_CSTRING(args[0]);
+    printf("%04d/%02d/%02d %02d:%02d:%02d %s", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+    
+    exit(1);
+}
+
+static Value fatallnLog(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "fatalln() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+    
+    struct tm *ptm = timeNow();
+
+    char *msg = AS_CSTRING(args[0]);
+    printf("%04d/%02d/%02d %02d:%02d:%02d %s\n", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+    
+    exit(1);
+}
+
+void freeLog(DictuVM *vm, ObjAbstract *abstract) {
+    Log *log = (Log*)abstract->data;
+    if (log->of != NULL) {
+        fclose(log->of);
+    }
+    
+    if (log->prefix != NULL) {
+        free(log->prefix);
+    }
+
+    FREE(vm, Log, abstract->data);
+}
+
+static Value logObjSetPrefix(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "setPrefix() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
+    char *prefix = AS_CSTRING(args[1]);
+    
+    if (strlen(prefix) == 0) {
+        return NIL_VAL;
+    }
+    
+    log->prefix = strdup(prefix);
+
+    return NIL_VAL;
+}
+
+static Value logObjUnsetPrefix(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "unsetPrefix() takes 0 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
```

Comment:
Will this leak memory? My initial feelings are we will need to free the old `log->prefix` if it's not NULL.